soft4pes/control/mpc/solvers/mpc_enum.py

```diff
@@ -0,0 +1,172 @@
+
+    Attributes
+    ----------
+    J : 1 x conv.nl^(3*Np) ndarray of floats
+        Array for costs.
+    u_seq : 3*Np x conv.nl^(3*Np) ndarray of ints
+        Array for 3-phase switch position sequences.
+    i : int
+        Counter for the switch position sequences.
+    SW_COMB : 1 x conv.nl^3 ndarray of ints
+        All possible 3-phase switch position.
+    """
+
+    def __init__(self, conv):
+        """
+        Initialize an MpcEnum instance.
+
+        Parameters
+        ----------
+        conv : converter object
+            Converter model.
+        """
+        self.J = None
+        self.u_seq = None
+        self.i = 0
+
+        if conv.nl == 2:
+            sw_pos_3ph = np.array([-1, 1])
+
+        elif conv.nl == 3:
+            sw_pos_3ph = np.array([-1, 0, 1])
+
+        # Create all possible 3-phase switch position
+        self.SW_COMB = np.array(list(product(sw_pos_3ph, repeat=3)))
+
+    def __call__(self, sys, conv, ctr, ref):
+        """
+        Solve MPC problem by using exhaustive enumeration.
+
+        Parameters
+        ----------
+        sys : system object
+            System model.
+        conv : converter object
+            Converter model.
+        ctr : controller object
+            Controller model.
+        ref : ndarray of floats
+            Reference.
+
+        Returns
+        -------
+        uk : 1 x 3 ndarray of ints
+            The 3-phase switch position sequence with the lowest cost.
+        """
+
+        # Initialize array for costs and 3-phase switch position sequences
+        self.J = np.zeros((conv.nl**(3 * ctr.Np), 1))
+        self.u_seq = np.zeros((conv.nl**(3 * ctr.Np), 3 * ctr.Np))
+        self.i = 0
+
+        self.solve(sys, conv, ctr, sys.x, ref, ctr.u_km1, 0)
+
+        # Find the 3-phase switch position sequence with the lowest cost
+        min_index = np.argmin(self.J)
+        uk = self.u_seq[min_index, 0:3]
+        return uk
+
+    def solve(self, sys, conv, ctr, xk, ref, u_km1, k):
+        """
+        Recursively compute the cost for different 3-phase switch position sequences.
+
+        Parameters
+        ----------
+        sys : system object
+            System model.
+        conv : converter object
+            Converter model.
+        xk : ndarray of floats
+            Current state vector.
+        ref : 1 x 2 ndarray of floats
+            Reference.
+        u_km1 : 1 x 3 ndarray of ints
+            Previous 3-phase switch position.
+        k : int
+            Prediction step.
+        """
+
+        # Calculate the range of columns for the current prediction step
+        # This is done in order to save calculation effort, as redundant calculations are avoided
+        u_col_range_start = 3 * k
+        u_col_range_end = 3 * (k + 1)
+        u_col_range = range(u_col_range_start, u_col_range_end)
+
+        # Make a rotation matrix
+        delta_theta = sys.wg * ctr.Ts * sys.base.w
```

Comment:
As before, let's have the normalized Ts on a different line.

I would also probably build the reference vector outside the recursive calls

Reply:
Solved in the next commit